### PR TITLE
Fixes push kidnapping with fast xenos 

### PR DESCRIPTION
**Pull request by MacBlaze1**
## About The Pull Request
This PR makes adds a cooldown on pushing for xenos so they dont lightspeed someone away
## Why It's Good For The Game

This allows for an end to xeno pushing 10 tiles away with a runner and then murdering the marine there. I was told to change this by many people. This also fixes annoying runners just bodyblocking higher tier xenos.

## Changelog
:cl: Blase, SplinterGP
balance: there is a 5 tick cooldown on pushing humans for xenos, preventing runners from carrying people away at lightspeed
/:cl:

### CODE REVIEW

File: code/modules/mob/living/living.dm [hunk 1]
```diff
@@ -386,6 +389,10 @@
 			return
 
 	if(ismovableatom(A))
+		if(isxeno(src))
+			var/mob/living/carbon/xenomorph/meanie = src
+			if(meanie.xeno_caste.caste_flags & CASTE_CANT_PUSH && (istype(A,/mob/living/carbon/human) || istype(A, /mob/living/carbon/xenomorph)) && (a_intent == INTENT_HARM || a_intent == INTENT_DISARM || a_intent == INTENT_GRAB))
```
Comment (Wayland-Smithy) [inline code/modules/mob/living/living.dm:394]: Are you sure about including `INTENT_DISARM` and `INTENT_GRAB`?

Going to be pretty annoying to toggle bump off just to bully hosts that break out of nests when they go running for an exit.

Comment (SplinterGP) [reply code/modules/mob/living/living.dm:394]: the problem is not even with bump, bump is slow and all, problem is that they when on not bump just literally zooms the human different than other xenos, ideally they should get more slowdown when pushing but, needs some other coding to do, i'll see if i can. also missing a parenthesis, will warn blase

Comment (Wayland-Smithy) [reply code/modules/mob/living/living.dm:394]: Queens literally can't push a human around no matter what now nor can any other caste not just `CASTE_CANT_PUSH ` ones. Is that the intent? Splinter says there may just be parentheses missing, but aside from the runner and hunter I still think xenos should be able to push humans.
G2ffepXx1k

Comment (Wayland-Smithy) [reply code/modules/mob/living/living.dm:394]: Also if we do remove pushing the bump attack button now lies to you when you toggle it off: `You will now push enemies who are in your way.`

Comment (SplinterGP) [inline code/modules/mob/living/living.dm:394]: should have a parent more unifying these two
so from
`(istype(A,/mob/living/carbon/human) || istype(A, /mob/living/carbon/xenomorph))`
to
`((istype(A,/mob/living/carbon/human) || istype(A, /mob/living/carbon/xenomorph)))`

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/living.dm (hunk 1 above)
Comment (SplinterGP) [reply code/modules/mob/living/living.dm:394]: yeah bug, also when bump is off should just be, "You will now not attack enemies on your way"

Comment (SplinterGP) [reply code/modules/mob/living/living.dm:394]: since on bump you push as well(but slower)

Comment (MacBlaze1) [reply code/modules/mob/living/living.dm]: done

File: code/modules/mob/living/living.dm [hunk 2]
```diff
@@ -386,6 +389,10 @@
 			return
 
 	if(ismovableatom(A))
+		if(isxeno(src))
+			var/mob/living/carbon/xenomorph/meanie = src
+			if(meanie.xeno_caste.caste_flags & CASTE_CANT_PUSH && ((istype(A,/mob/living/carbon/human) || istype(A, /mob/living/carbon/xenomorph))) && (a_intent == INTENT_HARM || a_intent == INTENT_DISARM || a_intent == INTENT_GRAB))
```
Comment (SplinterGP) [inline code/modules/mob/living/living.dm:394]:
```suggestion
			if((meanie.xeno_caste.caste_flags & CASTE_CANT_PUSH) && ((istype(A,/mob/living/carbon/human) || istype(A, /mob/living/carbon/xenomorph)))
```

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/living.dm (hunk 2 above)
Comment (SplinterGP) [reply code/modules/mob/living/living.dm]: will always reach on harm grab or disarm anyway, also missing a parenthesis

**Comment by SplinterGP** [comment]
wtf was this travis error, travis you dingus

File: code/modules/mob/living/living.dm [hunk 3]
```diff
@@ -353,6 +354,8 @@
 				mob_swap = TRUE
 			else if(mob_size > L.mob_size && a_intent == INTENT_HELP) //Larger mobs can shove aside smaller ones.
 				mob_swap = TRUE
+			else if((X.xeno_caste.caste_flags & CASTE_ALWAYS_PUSH_PAST) && isxeno(src))
+				mob_swap = TRUE
```
Comment (Wayland-Smithy) [inline code/modules/mob/living/living.dm:358]: I got bad news. Pushing a human as any xeno still does not work at all.

A little debugging showed me that `mob_swap = TRUE` is happening regardless of intent.

What if `L` and `X` as a result are not actually a xeno? I think that is causing the logic to break.

Comment (Wayland-Smithy) [reply code/modules/mob/living/living.dm:358]: -rep for not TESTING YOUR OWN SHIT MAGGOT
monkey+keyboard

Comment (MacBlaze1) [reply code/modules/mob/living/living.dm:358]: Ive been a little busy with like 4 other things

Comment (Wayland-Smithy) [reply code/modules/mob/living/living.dm:358]: I think we would all prefer delayed and tested code over speedy and untested. It is not like there is any pressure besides that which you put on yourself to get things pushed and PRed so quickly as to be error prone.

**Comment by Chaznoodles** [comment]
Should just make all small xenos unable to push, rather than just hunters/runners. They're not the only guilty parties.

**Comment by MacBlaze1** [comment]
> Should just make all small xenos unable to push, rather than just picking and choosing. The ones you've picked aren't the only guilty parties.

there are only big xenos, and normal xenos

File: code/__DEFINES/mobs.dm [hunk 4]
```diff
@@ -459,6 +459,7 @@ GLOBAL_LIST_INIT(xenoupgradetiers, list(XENO_UPGRADE_BASETYPE, XENO_UPGRADE_INVA
 #define CASTE_ACID_BLOOD (1<<13) //The acid blood effect which damages humans near xenos that take damage
 #define CASTE_CAN_HOLD_JELLY (1<<14)//whether we can hold fireproof jelly in our hands
 
+
 //Charge-Crush
 #define CHARGE_OFF			0
```
Comment (TiviPlus) [inline code/__DEFINES/mobs.dm:464]: fix

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 5]
```diff
@@ -123,3 +123,6 @@
 
 	///Amount of deciseconds gained from the braindeath timer, usually by CPR.
 	var/revive_grace_time = 0
+
+	var/next_push = 0 // Time of next push
+	var/push_delay = 5 // push delay
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/human_defines.dm:128]: Use the cooldown defines in cooldowns.dm and make this into a define instead

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 6]
```diff
@@ -244,3 +244,4 @@
 	var/notice_delay = 20 //2 second between notices
 
 	var/fire_luminosity = 0 //Luminosity of the current fire while burning
+
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:247]: fix

File: code/modules/mob/living/living.dm [hunk 7]
```diff
@@ -345,6 +345,7 @@
 
 		if(!L.buckled && !L.anchored)
 			var/mob_swap = FALSE
+
 			//the puller can always swap with its victim if on grab intent
```
Comment (TiviPlus) [inline code/modules/mob/living/living.dm:349]: fix

File: code/modules/mob/living/living.dm [hunk 8]
```diff
@@ -386,6 +387,13 @@
 			return
 
 	if(ismovableatom(A))
+		if(isxeno(src))
+			if(ishuman(A))
+				var/mob/living/carbon/human/H = A
+				if(H.next_push > world.time)
+					return
+				else
+					H.next_push = world.time + H.push_delay
```
Comment (TiviPlus) [inline code/modules/mob/living/living.dm:396]:
```suggestion
		if(isxeno(src) && ishuman(A))
			var/mob/living/carbon/human/H = A
			if(H.next_push > world.time)
				return
			else
				H.next_push = world.time + H.push_delay
```
And use the defines

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Also fix your CL to detail what it does

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Comment by davidchan13** [comment]
Hivelords using resin walker make this a huge pain in the ass, since they can basically keep pace with a crusher if they are on weeds.

File: code/__DEFINES/cooldowns.dm [hunk 9]
```diff
@@ -24,7 +24,7 @@
 #define COOLDOWN_CPR		"CPR"
 #define COOLDOWN_IV_PING	"iv_ping"
 #define COOLDOWN_RACK_BOLT	"rack_bolt"
-
+#define COOLDOWN_PUSH_DELAY "push_delay"
 //// COOLDOWN SYSTEMS
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/__DEFINES/cooldowns.dm:28]: stop self resolving change requests

[the remark this answers is not in the document]
Comment (MacBlaze1) [reply code/__DEFINES/cooldowns.dm]: did this

File: code/modules/mob/living/living.dm [hunk 10]
```diff
@@ -386,6 +386,12 @@
 			return
 
 	if(ismovableatom(A))
+		if(isxeno(src) && ishuman(A))
+			var/mob/living/carbon/human/H = A
+			if(TIMER_COOLDOWN_CHECK(H, COOLDOWN_PUSH_DELAY))
+				return
+			else
+				TIMER_COOLDOWN_START(H, COOLDOWN_PUSH_DELAY, 5)
 		PushAM(A)
```
[the remark this answers is not in the document]
Comment (MacBlaze1) [reply code/modules/mob/living/living.dm]: did this too

File: code/__DEFINES/mobs.dm [hunk 11]
```diff
@@ -190,6 +190,7 @@ GLOBAL_LIST_INIT(xenoupgradetiers, list(XENO_UPGRADE_BASETYPE, XENO_UPGRADE_INVA
 /////////////////MOVE DEFINES//////////////////////
 #define MOVE_INTENT_WALK        0
 #define MOVE_INTENT_RUN         1
+#define HUMAN_PUSHED_DELAY      5
 ///////////////////INTERNAL ORGANS DEFINES///////////////////
```
Comment (TiviPlus) [inline code/__DEFINES/mobs.dm:194]:
```suggestion
#define MOVE_INTENT_WALK        0
#define MOVE_INTENT_RUN         1
#define HUMAN_PUSHED_DELAY      5

///////////////////INTERNAL ORGANS DEFINES///////////////////
```

File: code/__DEFINES/cooldowns.dm [hunk 12]
```diff
@@ -24,7 +24,6 @@
 #define COOLDOWN_CPR		"CPR"
 #define COOLDOWN_IV_PING	"iv_ping"
 #define COOLDOWN_RACK_BOLT	"rack_bolt"
-
 //// COOLDOWN SYSTEMS
```
Comment (TiviPlus) [inline code/__DEFINES/cooldowns.dm:27]:
```suggestion

//// COOLDOWN SYSTEMS
```

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 13]
```diff
@@ -123,3 +123,5 @@
 
 	///Amount of deciseconds gained from the braindeath timer, usually by CPR.
 	var/revive_grace_time = 0
+	COOLDOWN_DECLARE(push_delay)
+
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/human_defines.dm:127]:
```suggestion
	var/revive_grace_time = 0

	COOLDOWN_DECLARE(push_delay)

```

File: code/modules/mob/living/living.dm [hunk 14]
```diff
@@ -386,6 +386,12 @@
 			return
 
 	if(ismovableatom(A))
+		if(isxeno(src) && ishuman(A))
+			var/mob/living/carbon/human/H = A
+			if(!COOLDOWN_CHECK(H,  push_delay))
+				return
+			else
+				COOLDOWN_START(H, push_delay, HUMAN_PUSHED_DELAY)
```
Comment (TiviPlus) [inline code/modules/mob/living/living.dm:394]:
```suggestion
			COOLDOWN_START(H, push_delay, HUMAN_PUSHED_DELAY)
```

**Review by TiviPlus — CHANGES_REQUESTED** [review]
rename HUMAN_PUSHED_DELAY to something like XENO_HUMAN_PUSH_DELAY

and push_delay to something like xeno_push_delay

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**